### ISSUE TITLE
Metron high cpu load

### ISSUE THREAD

**Issue by deltatree** [cloudfoundry/cf-release#1200]
Hi, I'm testing release 259 - The metron agent is consuming a lot of cpu - it is slowing down all. For deploying I am using: https://github.com/cloudfoundry/bosh-deployment with the default configuration.
Someone else with these problems?
Maybe I am adding this to the wrong repo?? ...
Thanks a lot

Alex

**Comment by ahevenor** [comment]
@deltatree Are you on a bosh-lite? Do you have v2 API enabled for metron? The property is ` loggregator.use_v2_api` https://github.com/cloudfoundry/diego-release/blob/fe795cfa4b9daa2bbd7b6b68df8dccb187338314/jobs/rep/spec#L77.

You can likely set this to false to mitigate the issue. Before you do that however would you mind providing us with a cpu profile? Here are the steps to get a profile:

1. ssh on to the offending instance.
1. `sudo -1`
1. Find the pprof port: `lsof -c metron | grep TCP | grep LISTEN | grep -v 3457 | grep -v 3458 | grep -v 9090 | grep -v 8080` That should list the port pprof is listening on.
1. Get a cpu profile: `curl http://localhost:41117/debug/pprof/profile?debug=1 > profile.out`
1. Once you have that file you can send it to us on slack [#loggregator](https://cloudfoundry.slack.com/messages/C02HCCXV5)

**Comment by deltatree** [comment]
Yes I am on bosh-lite. Loggregator was on default - I changed it but there is no improvement.

I've attached the cpu profile from one instance:
[profile.zip](https://github.com/cloudfoundry/cf-release/files/1002650/profile.zip)

All instances have the same behaviour.

Thanks a lot

Alex

**Comment by ahevenor** [comment]
@deltatree - I am surprised switching off the v2 flag didn't resolve this. There is a known CPU issue with v2 that we are working to correct. That said we don't see anything obvious in the profile. When you see this again can you please contact us in [#loggregator](https://pivotal.slack.com/messages/C3MGZ8FP1) so we can help troubleshoot.

**Comment by deltatree** [comment]
@ahevenor - Hi, how can I login to slack? The only channel I currently have is this here...

What do you mean with again? It is all the time - have a look:
![bosh-lite-vm-htop](https://cloud.githubusercontent.com/assets/6955880/26145295/39a68a9e-3aec-11e7-9b02-8ced8f5be7db.png)

**Comment by ahevenor** [comment]
We introduced an improvement which batches messages inbound to metron that should help with CPU Utilization. This was released in Loggregator 89 and is expected to be included in CF 263 (currently in release election process). I am not sure this address the scenario @deltatree is seeing but it improves CPU utilization in our tests.

**Comment by dsabeti** [comment]
Hi all. Any updates on this? @deltatree, let us know if the loggregator updates have helped with this.

**Comment by dsabeti** [comment]
I'm going to close this issue due to inactivity, but feel free to re-open the issue if you need more help.

**Comment by deltatree** [comment]
Hi, sorry for commenting so late: I was on holiday.
I tried with current versions, but with the default setup I'm getting a very high load ... if I look into the vms: metron is consuming all of the cpu...

![high-cpu-after-new-deployment](https://user-images.githubusercontent.com/6955880/27214611-a8ae06ea-526d-11e7-992f-6cadddb049ca.png)

bosh2 -e vbox releases
Using environment '192.168.50.6' as client 'admin'

Name                     Version   Commit Hash
binary-buildpack         1.0.13*   0bde2c77+
capi                     1.30.0*   f930be1a
cf-mysql                 35        4947ee5
cf-networking            1.0.0     560e00f
cf-smoke-tests           26*       a2b4d73
cflinuxfs2               1.130.0*  4f8bf79d+
consul                   170*      06fd84c
datadog-firehose-nozzle  59        5bd8b794+
diego                    1.18.1*   58757374
dotnet-core-buildpack    1.0.19*   cd3bbb05+
etcd                     112*      c97b2c3
garden-runc              1.8.0*    16f64c91
garden-windows           0.4.0     942bd2a6
go-buildpack             1.8.4*    c29b5300+
grootfs                  0.19.0    53831be
hwc-buildpack            2.3.4     cdb29cf7+
java-buildpack           3.17*     298f4bd5+
loggregator              89*       e5611f8
nats                     17*       faf1d99e+
nodejs-buildpack         1.5.36*   ecd44abe+
php-buildpack            4.3.34*   c834c1cc+
postgres                 17        b21260ee+
~                        16*       e3cfa2b2+
python-buildpack         1.5.19*   6e5bc32a+
routing                  0.157.0*  114fe5e9
ruby-buildpack           1.6.40*   060529d1+
staticfile-buildpack     1.4.6*    e5ea4334+
statsd-injector          1.0.28*   4c0ffe4
uaa                      41*       83b20714

(*) Currently deployed
(+) Uncommitted changes

29 releases

**Comment by dsabeti** [comment]
Ok, if you're still seeing this issue, then maybe @ahevenor (cc @jasonkeene) can help out. 

Like Adam said, you can also ask for help in the #loggregator channel of the Cloud Foundry slack.
Cloud Foundry slack: https://cloudfoundry.slack.com/
#loggregator channel: https://cloudfoundry.slack.com/messages/C02HCCXV5/

**Comment by markikollasch** [comment]
We've observed this issue recently on our end as well, and hypothesized that the high CPU usage across VMs was being exacerbated by the ordering of the deployment. By moving `doppler` to deploy immediately after `etcd`, each job with a `metron` component should be able to successfully establish a connection to it that much sooner, and consume fewer resources by repeatedly failing. We tested this and found improvement, so we've opened [PR 165 on `cloudfoundry/cf-deployment`](https://github.com/cloudfoundry/cf-deployment/pull/165).
